exporters/otlp/otlp.go

```diff
@@ -0,0 +1,257 @@
+	var err = errAlreadyStarted
+	e.startOnce.Do(func() {
+		e.mu.Lock()
+		e.started = true
+		e.disconnectedCh = make(chan bool, 1)
+		e.stopCh = make(chan bool)
+		e.backgroundConnectionDoneCh = make(chan bool)
+		e.mu.Unlock()
+
+		// An optimistic first connection attempt to ensure that
+		// applications under heavy load can immediately process
+		// data. See https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/pull/63
+		if err := e.connect(); err == nil {
+			e.setStateConnected()
+		} else {
+			e.setStateDisconnected(err)
+		}
+		go e.indefiniteBackgroundConnection()
+
+		err = nil
+	})
+
+	return err
+}
+
+func (e *Exporter) prepareCollectorAddress() string {
+	if e.c.collectorAddr != "" {
+		return e.c.collectorAddr
+	}
+	return fmt.Sprintf("%s:%d", DefaultCollectorHost, DefaultCollectorPort)
+}
+
+func (e *Exporter) enableConnections(cc *grpc.ClientConn) error {
+	e.mu.RLock()
+	started := e.started
+	e.mu.RUnlock()
+
+	if !started {
+		return errNotStarted
+	}
+
+	e.mu.Lock()
+	// If previous clientConn is same as the current then just return.
+	// This doesn't happen right now as this func is only called with new ClientConn.
+	// It is more about future-proofing.
+	if e.grpcClientConn == cc {
+		e.mu.Unlock()
+		return nil
+	}
+	// If the previous clientConn was non-nil, close it
+	if e.grpcClientConn != nil {
+		_ = e.grpcClientConn.Close()
+	}
+	e.grpcClientConn = cc
+	e.traceExporter = coltracepb.NewTraceServiceClient(cc)
+	e.mu.Unlock()
+
+	return nil
+}
+
+func (e *Exporter) dialToCollector() (*grpc.ClientConn, error) {
+	addr := e.prepareCollectorAddress()
+	var dialOpts []grpc.DialOption
+	if e.c.clientCredentials != nil {
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(e.c.clientCredentials))
+	} else if e.c.canDialInsecure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+	if e.c.compressor != "" {
+		dialOpts = append(dialOpts, grpc.WithDefaultCallOptions(grpc.UseCompressor(e.c.compressor)))
+	}
+	if len(e.c.grpcDialOptions) != 0 {
+		dialOpts = append(dialOpts, e.c.grpcDialOptions...)
+	}
+
+	ctx := context.Background()
+	if len(e.c.headers) > 0 {
+		ctx = metadata.NewOutgoingContext(ctx, metadata.New(e.c.headers))
+	}
+	return grpc.DialContext(ctx, addr, dialOpts...)
+}
+
+// Stop shuts down all the connections and resources
+// related to the exporter.
+// If the exporter is not started then this func does nothing.
+func (e *Exporter) Stop() error {
+	e.mu.RLock()
+	cc := e.grpcClientConn
+	started := e.started
+	e.mu.RUnlock()
+
+	if !started {
+		return nil
+	}
+
+	// Now close the underlying gRPC connection.
+	var err error
+	if cc != nil {
+		err = cc.Close()
+	}
+
+	// At this point we can change the state variable started
+	e.mu.Lock()
+	e.started = false
+	e.mu.Unlock()
+	close(e.stopCh)
+
+	// Ensure that the backgroundConnector returns
+	<-e.backgroundConnectionDoneCh
+
+	return err
+}
+
+func (e *Exporter) ExportSpans(ctx context.Context, sds []*export.SpanData) {
```

Comment:
I'm wondering if we should pass `ctx` down to the `uploadTraces` function, so the caller of the `ExportSpans` function can decide about the timeout to use instead of us hardcoding it to 1 minute.

Reply:
ExportSpans is called by the BatchSpanProcessor. So is the value of timeout an option for the Batcher irrespective of which exporter it is exporting to or is it the option for the exporter. I think it would the latter. We can expose a new config option for the exporter.